Improve FluidSynth support

Continuation of #262

During #262 we identified few areas that need improvements - especially with the goal of supporting MIDI emulation on lower-specced computers like Raspberry Pi with possibly slow storage via an SD card.

Needed improvements:

1. Putting fluidsynth synthesis on a separate thread. Right now synthesis thread is shared with emulator (which is fast enough for handling synth even on older PCs, but fluidsynth complains about it and it slows down emulation). This work might require quite severe improvements to DOSBox mixer implementation, but this should pay dividends when integrating MT-32.
2. FluidSynth library has a quite extensive dependency tree via glib - they have some important reasons for doing so (e.g. POSIX is missing high-resolution timers required). On Linux it's not a big issue, but it causes problems with packaging for Windows and macOS. We should cooperate with FluidSynth team to find a way to replace those dependencies (it might be possible e.g. via C++11).
3. Slow storage mediums (slow HDDs or SD cards) when used with medium-to-large soundfonts can cause synthesis issues when emulator performs disk access - perhaps placing fluidsynth initialization in a separate thread can prevent this.
4. SoundFont loading on slows down the emulator startup (visibly - by ~0.1-0.2s) - maybe solution to (1) or (3) will solve this problem, but we need to consider improving splash screen implementation, so the soundfont loading would happen in parallel.

If anyone has spare time to tackle any of these problems: feel free to start working on them! :)

**Migrate FluidSynth calls to non-deprecated versions**

``` text
 midi_fluidsynth.cpp:240:2: warning: 'fluid_synth_set_chorus_on' is deprecated [-Wdeprecated-declarations]
        fluid_synth_set_chorus_on(fluid_synth.get(), 1);
        ^
/Users/runner/work/dosbox-staging/dosbox-staging/contrib/static-fluidsynth/include/fluidsynth/synth.h:208:1: note: 'fluid_synth_set_chorus_on' has been explicitly marked deprecated here
FLUID_DEPRECATED FLUIDSYNTH_API void fluid_synth_set_chorus_on(fluid_synth_t *synth, int on);
^
/Users/runner/work/dosbox-staging/dosbox-staging/contrib/static-fluidsynth/include/fluidsynth.h:58:45: note: expanded from macro 'FLUID_DEPRECATED'
#    define FLUID_DEPRECATED __attribute__((deprecated))
                                            ^
midi_fluidsynth.cpp:241:2: warning: 'fluid_synth_set_chorus' is deprecated [-Wdeprecated-declarations]
        fluid_synth_set_chorus(fluid_synth.get(), chorus_number, chorus_level,
        ^
/Users/runner/work/dosbox-staging/dosbox-staging/contrib/static-fluidsynth/include/fluidsynth/synth.h:211:1: note: 'fluid_synth_set_chorus' has been explicitly marked deprecated here
FLUID_DEPRECATED FLUIDSYNTH_API int fluid_synth_set_chorus(fluid_synth_t *synth, int nr, double level,
^
/Users/runner/work/dosbox-staging/dosbox-staging/contrib/static-fluidsynth/include/fluidsynth.h:58:45: note: expanded from macro 'FLUID_DEPRECATED'
#    define FLUID_DEPRECATED __attribute__((deprecated))
                                            ^
midi_fluidsynth.cpp:248:2: warning: 'fluid_synth_set_reverb_on' is deprecated [-Wdeprecated-declarations]
        fluid_synth_set_reverb_on(fluid_synth.get(), 1);
        ^
/Users/runner/work/dosbox-staging/dosbox-staging/contrib/static-fluidsynth/include/fluidsynth/synth.h:162:1: note: 'fluid_synth_set_reverb_on' has been explicitly marked deprecated here
FLUID_DEPRECATED FLUIDSYNTH_API void fluid_synth_set_reverb_on(fluid_synth_t *synth, int on);
^
/Users/runner/work/dosbox-staging/dosbox-staging/contrib/static-fluidsynth/include/fluidsynth.h:58:45: note: expanded from macro 'FLUID_DEPRECATED'
#    define FLUID_DEPRECATED __attribute__((deprecated))
                                            ^
midi_fluidsynth.cpp:249:2: warning: 'fluid_synth_set_reverb' is deprecated [-Wdeprecated-declarations]
        fluid_synth_set_reverb(fluid_synth.get(), reverb_room_size,
        ^
/Users/runner/work/dosbox-staging/dosbox-staging/contrib/static-fluidsynth/include/fluidsynth/synth.h:165:1: note: 'fluid_synth_set_reverb' has been explicitly marked deprecated here
FLUID_DEPRECATED FLUIDSYNTH_API int fluid_synth_set_reverb(fluid_synth_t *synth, double roomsize,
^
/Users/runner/work/dosbox-staging/dosbox-staging/contrib/static-fluidsynth/include/fluidsynth.h:58:45: note: expanded from macro 'FLUID_DEPRECATED'
#    define FLUID_DEPRECATED __attribute__((deprecated))
```

--- a/.github/workflows/analysis.yml
+++ b/.github/workflows/analysis.yml
@@ -82,7 +82,7 @@ jobs:
           path: report
       - name: Summarize report
         env:
-          MAX_BUGS: 183
+          MAX_BUGS: 184
         run: |
           # summary
           echo "Full report is included in build Artifacts"

--- a/src/midi/midi_fluidsynth.cpp
+++ b/src/midi/midi_fluidsynth.cpp
@@ -240,11 +240,8 @@ bool MidiHandlerFluidsynth::Open(MAYBE_UNUSED const char *conf)
 	// Let the user know that the SoundFont was loaded
 	if (scale_by_percent == 100)
 		LOG_MSG("MIDI: Using SoundFont '%s'", soundfont.c_str());
-	else if (scale_by_percent > 100)
-		LOG_MSG("MIDI: Using SoundFont '%s' with levels amplified by %d%%",
-		        soundfont.c_str(), scale_by_percent);
 	else
-		LOG_MSG("MIDI: Using SoundFont '%s' with levels attenuated by %d%%",
+		LOG_MSG("MIDI: Using SoundFont '%s' with voices scaled by %d%%",
 		        soundfont.c_str(), scale_by_percent);
 
 	constexpr int fx_group = -1; // applies setting to all groups

--- a/src/midi/midi_fluidsynth.cpp
+++ b/src/midi/midi_fluidsynth.cpp
@@ -28,11 +28,12 @@
 #include <deque>
 #include <string>
 #include <tuple>
-#include <vector>
 
 #include "control.h"
 #include "cross.h"
 #include "fs_utils.h"
+
+static constexpr int FRAMES_PER_BUFFER = 512; // synth granularity
 
 MidiHandlerFluidsynth instance;
 
@@ -41,7 +42,7 @@ static void init_fluid_dosbox_settings(Section_prop &secprop)
 	constexpr auto when_idle = Property::Changeable::WhenIdle;
 
 	// Name 'default.sf2' picks the default soundfont if it's installed
-	// it the OS. Usually it's Fluid_R3.
+	// in the OS. Usually it's Fluid_R3.
 	auto *str_prop = secprop.Add_string("soundfont", when_idle, "default.sf2");
 	str_prop->Set_help(
 	        "Path to a SoundFont file in .sf2 format. You can use an\n"
@@ -51,13 +52,6 @@ static void init_fluid_dosbox_settings(Section_prop &secprop)
 	        "An optional percentage will scale the SoundFont's volume.\n"
 	        "For example: 'soundfont.sf2 50' will attenuate it by 50 percent.\n"
 	        "The scaling percentage can range from 1 to 500.");
-
-	auto *int_prop = secprop.Add_int("synth_threads", when_idle, 1);
-	int_prop->SetMinMax(1, 256);
-	int_prop->Set_help(
-	        "If set to a value greater than 1, then additional synthesis\n"
-	        "threads will be created to take advantage of many CPU cores.\n"
-	        "(min 1, max 256)");
 }
 
 // SetMixerLevel is a callback that's given the user-desired mixer level,
@@ -180,7 +174,7 @@ static std::string find_sf_file(const std::string &name)
 }
 
 MidiHandlerFluidsynth::MidiHandlerFluidsynth()
-        : soft_limiter("FSYNTH", prescale_level, expected_max_frames)
+        : soft_limiter("FSYNTH", prescale_level, FRAMES_PER_BUFFER)
 {}
 
 bool MidiHandlerFluidsynth::Open(MAYBE_UNUSED const char *conf)
@@ -193,22 +187,23 @@ bool MidiHandlerFluidsynth::Open(MAYBE_UNUSED const char *conf)
 		LOG_MSG("MIDI: new_fluid_settings failed");
 		return false;
 	}
-	auto *section = static_cast<Section_prop *>(control->GetSection("fluidsynth"));
+
+	// Setup the mixer channel and level callback
+	const auto mixer_callback = std::bind(&MidiHandlerFluidsynth::MixerCallBack,
+	                                      this, std::placeholders::_1);
+	channel_t mixer_channel(MIXER_AddChannel(mixer_callback, 0, "FSYNTH"),
+	                        MIXER_DelChannel);
+
+	const auto set_mixer_level = std::bind(&MidiHandlerFluidsynth::SetMixerLevel,
+	                                       this, std::placeholders::_1);
+	mixer_channel->RegisterLevelCallBack(set_mixer_level);
 
 	// Detailed explanation of all available FluidSynth settings:
 	// http://www.fluidsynth.org/api/fluidsettings.xml
 
-	const int cpu_cores = section->Get_int("synth_threads");
-	fluid_settings_setint(fluid_settings.get(), "synth.cpu-cores", cpu_cores);
-
-	const auto mixer_callback = std::bind(&MidiHandlerFluidsynth::MixerCallBack,
-	                                      this, std::placeholders::_1);
-	mixer_channel_ptr_t mixer_channel(MIXER_AddChannel(mixer_callback, 0, "FSYNTH"),
-	                                  MIXER_DelChannel);
-
-	// Per the FluidSynth API, the sample-rate should be part of the settings used to
-	// instantiate the synth, so we create the mixer channel first and use its native
-	// rate to configure FluidSynth.
+	// Per the FluidSynth API, the sample-rate should be part of the
+	// settings used to instantiate the synth, so we create the mixer
+	// channel first and use its native rate to configure FluidSynth.
 	fluid_settings_setnum(fluid_settings.get(), "synth.sample-rate",
 	                      mixer_channel->GetSampleRate());
 
@@ -220,6 +215,7 @@ bool MidiHandlerFluidsynth::Open(MAYBE_UNUSED const char *conf)
 	}
 
 	// Load the requested SoundFont or quit if none provided
+	auto *section = static_cast<Section_prop *>(control->GetSection("fluidsynth"));
 	const auto sf_spec = parse_sf_pref(section->Get_string("soundfont"), 100);
 	const auto soundfont = find_sf_file(std::get<std::string>(sf_spec));
 	auto scale_by_percent = std::get<int>(sf_spec);
@@ -294,17 +290,25 @@ bool MidiHandlerFluidsynth::Open(MAYBE_UNUSED const char *conf)
 	                       reverb_damping, reverb_width, reverb_level);
 #endif
 
-	// Let the mixer command adjust our internal level
-	const auto set_mixer_level = std::bind(&MidiHandlerFluidsynth::SetMixerLevel,
-	                                       this, std::placeholders::_1);
-	mixer_channel->RegisterLevelCallBack(set_mixer_level);
-	mixer_channel->Enable(true);
-
 	settings = std::move(fluid_settings);
 	synth = std::move(fluid_synth);
 	channel = std::move(mixer_channel);
+
+	// Start rendering audio
+	keep_rendering = true;
+	const auto render = std::bind(&MidiHandlerFluidsynth::Render, this);
+	renderer = std::thread(render);
+	ring.wait_dequeue(play_buffer); // populate the first play buffer
+
+	// Start playback
+	channel->Enable(true);
 	is_open = true;
 	return true;
+}
+
+MidiHandlerFluidsynth::~MidiHandlerFluidsynth()
+{
+	Close();
 }
 
 void MidiHandlerFluidsynth::Close()
@@ -312,10 +316,25 @@ void MidiHandlerFluidsynth::Close()
 	if (!is_open)
 		return;
 
-	channel->Enable(false);
-	channel = nullptr;
-	synth = nullptr;
-	settings = nullptr;
+	// Stop playback
+	if (channel)
+		channel->Enable(false);
+
+	// Stop rendering and drain the ring
+	keep_rendering = false;
+	while (ring.size_approx())
+		ring.wait_dequeue(play_buffer);
+
+	// Wait for the rendering thread to finish
+	if (renderer.joinable())
+		renderer.join();
+
+	// Reset the members
+	channel.reset();
+	synth.reset();
+	settings.reset();
+	last_played_frame = 0;
+
 	is_open = false;
 }
 
@@ -375,17 +394,47 @@ void MidiHandlerFluidsynth::PrintStats()
 
 void MidiHandlerFluidsynth::MixerCallBack(uint16_t requested_frames)
 {
-	constexpr uint16_t max_frames = expected_max_frames; // local fixes link error
-	std::vector<float> render_buffer(max_frames * 2);    // L & R channels
+	while (requested_frames) {
+		const auto frames_to_be_played = std::min(GetRemainingFrames(),
+		                                          requested_frames);
+		const auto sample_offset_in_buffer = play_buffer.data() +
+		                                     last_played_frame * 2;
 
-	while (requested_frames > 0) {
-		const uint16_t frames = std::min(requested_frames, max_frames);
+		assert(frames_to_be_played <= play_buffer.size());
+		channel->AddSamples_s16(frames_to_be_played, sample_offset_in_buffer);
 
-		fluid_synth_write_float(synth.get(), frames, render_buffer.data(),
-		                        0, 2, render_buffer.data(), 1, 2);
-		const auto out = soft_limiter.Process(render_buffer, frames);
-		channel->AddSamples_s16(frames, out.data());
-		requested_frames -= frames;
+		requested_frames -= frames_to_be_played;
+		last_played_frame += frames_to_be_played;
+	}
+}
+
+// GetRemainingFrames() gives us the number of unplayed (ie: remaining) frames
+// left in the current play buffer. If it's run out, then we pop a new "full"
+// play buffer out of the ring. If the ring doesn't have any  to offer, then we
+// wait until it does.
+uint16_t MidiHandlerFluidsynth::GetRemainingFrames()
+{
+	// the current play buffer has some frames left, so return those
+	if (last_played_frame < FRAMES_PER_BUFFER)
+		return FRAMES_PER_BUFFER - last_played_frame;
+
+	// Otherwise move the next buffer out of the ring into our play buffer
+	ring.wait_dequeue(play_buffer);
+	last_played_frame = 0;
+
+	return FRAMES_PER_BUFFER;
+}
+
+// After GetRemainingFrames() pops out a buffer for playing, back-pressure is
+// released in the ring allowing FluidSynth to renderer the next buffer.
+void MidiHandlerFluidsynth::Render()
+{
+	std::vector<float> render_buffer(FRAMES_PER_BUFFER * 2); // L & R channels
+	while (keep_rendering) {
+		fluid_synth_write_float(synth.get(), FRAMES_PER_BUFFER,
+		                        render_buffer.data(), 0, 2, render_buffer.data(), 1, 2);
+		auto out = soft_limiter.Process(render_buffer, FRAMES_PER_BUFFER);
+		ring.wait_enqueue(out);
 	}
 }
 

--- a/src/midi/midi_fluidsynth.cpp
+++ b/src/midi/midi_fluidsynth.cpp
@@ -1,8 +1,8 @@
 /*
  *  SPDX-License-Identifier: GPL-2.0-or-later
  *
- *  Copyright (C) 2020-2021  Nikos Chantziaras <realnc@gmail.com>
  *  Copyright (C) 2020-2021  The DOSBox Staging Team
+ *  Copyright (C) 2020-2020  Nikos Chantziaras <realnc@gmail.com>
  *  Copyright (C) 2002-2011  The DOSBox Team
  *
  *  This program is free software; you can redistribute it and/or modify

--- a/src/midi/midi_fluidsynth.h
+++ b/src/midi/midi_fluidsynth.h
@@ -30,22 +30,16 @@
 #include <memory>
 #include <vector>
 #include <fluidsynth.h>
+#include <thread>
 
 #include "mixer.h"
+#include "../libs/rwqueue/readerwritercircularbuffer.h"
 #include "soft_limiter.h"
 
 class MidiHandlerFluidsynth final : public MidiHandler {
-private:
-	using fsynth_ptr_t = std::unique_ptr<fluid_synth_t, decltype(&delete_fluid_synth)>;
-
-	using fluid_settings_ptr_t =
-	        std::unique_ptr<fluid_settings_t, decltype(&delete_fluid_settings)>;
-
-	using mixer_channel_ptr_t =
-	        std::unique_ptr<MixerChannel, decltype(&MIXER_DelChannel)>;
-
 public:
 	MidiHandlerFluidsynth();
+	~MidiHandlerFluidsynth() override;
 	void PrintStats();
 	const char *GetName() const override { return "fluidsynth"; }
 	bool Open(const char *conf) override;
@@ -54,16 +48,29 @@ public:
 	void PlaySysex(uint8_t *sysex, size_t len) override;
 
 private:
-	static constexpr uint16_t expected_max_frames = (96000 / 1000) + 4;
 	void MixerCallBack(uint16_t requested_frames);
 	void SetMixerLevel(const AudioFrame &prescale_level) noexcept;
+	uint16_t GetRemainingFrames();
+	void Render();
+
+	using fluid_settings_ptr_t =
+	        std::unique_ptr<fluid_settings_t, decltype(&delete_fluid_settings)>;
+	using fsynth_ptr_t = std::unique_ptr<fluid_synth_t, decltype(&delete_fluid_synth)>;
+	using channel_t = std::unique_ptr<MixerChannel, decltype(&MIXER_DelChannel)>;
+	using ring_t = moodycamel::BlockingReaderWriterCircularBuffer<std::vector<int16_t>>;
+	using conditional_t = moodycamel::weak_atomic<bool>;
 
 	fluid_settings_ptr_t settings{nullptr, &delete_fluid_settings};
 	fsynth_ptr_t synth{nullptr, &delete_fluid_synth};
-	mixer_channel_ptr_t channel{nullptr, MIXER_DelChannel};
+	channel_t channel{nullptr, MIXER_DelChannel};
+	std::vector<int16_t> play_buffer = {};
+	ring_t ring{8}; // Handle up to eight buffers in the ring
+	std::thread renderer = {};
 	AudioFrame prescale_level = {1.0f, 1.0f};
 	SoftLimiter soft_limiter;
 
+	uint16_t last_played_frame = 0; // relative frame-offset in the play buffer
+	conditional_t keep_rendering = false;
 	bool is_open = false;
 };
 

--- a/src/midi/midi_fluidsynth.h
+++ b/src/midi/midi_fluidsynth.h
@@ -1,8 +1,8 @@
 /*
  *  SPDX-License-Identifier: GPL-2.0-or-later
  *
- *  Copyright (C) 2020-2021  Nikos Chantziaras <realnc@gmail.com>
  *  Copyright (C) 2020-2021  The DOSBox Staging Team
+ *  Copyright (C) 2020-2020  Nikos Chantziaras <realnc@gmail.com>
  *  Copyright (C) 2002-2011  The DOSBox Team
  *
  *  This program is free software; you can redistribute it and/or modify

